packages/block-editor/src/store/reducer.js

```diff
@@ -716,6 +720,132 @@ const withSaveReusableBlock = ( reducer ) => ( state, action ) => {
 	return reducer( state, action );
 };
 
+function aliasBlockTree( state, rootClientId, oldAliasState ) {
+	const aliasState = { ...oldAliasState };
+
+	const doAliasing = ( _rootClientId ) => {
+		return map( state.order[ _rootClientId ], ( clientId ) => {
+			const block = state.byClientId[ clientId ];
+			if ( ! block ) {
+				return null;
+			}
+
+			// Modify alias state along the way to keep everything organized.
+			const newId = uuid();
+			aliasState.aliasesOf[ newId ] = clientId;
+			if ( ! Array.isArray( aliasState.aliases[ clientId ] ) ) {
+				aliasState.aliases[ clientId ] = [];
+			}
+			aliasState.aliases[ clientId ].push( newId );
+
+			return {
+				...block,
+				attributes: state.attributes[ clientId ],
+				innerBlocks: !! state.controlledInnerBlocks[ clientId ]
+					? EMPTY_ARRAY
+					: doAliasing( clientId ),
+				clientId: newId,
+			};
+		} );
+	};
+
+	// Return the new information:
+	const newBlockTree = doAliasing( rootClientId );
+	return {
+		aliasState,
+		newBlockTree,
+	};
+}
+
+function getAliasedIds( state, clientId ) {
+	const isSource = !! state.aliases.aliases[ clientId ]; // has mapping source -> aliases
+	const isAlias = !! state.aliases.aliasesOf[ clientId ]; // has mapping alias -> source
+	if ( ! isSource && ! isAlias ) {
+		return null;
+	}
+	const sourceId = isSource ? clientId : state.aliases.aliasesOf[ clientId ];
+
+	// TODO: do I need to return the sourceID as well to include it in the chained action dispatch?
+	return [ ...state.aliases.aliases[ sourceId ], sourceId ];
+}
+
+// Actions which would otherwise be aliased, but should not be. These are mostly
+// UI-related, where we want the UI to show up for the specific block, not for
+// all blocks of the same alias.
+const actionsToIgnore = [
+	'SHOW_INSERTION_POINT',
+	'SET_INSERTION_POINT',
+	'SELECT_BLOCK',
+	'START_DRAGGING_BLOCKS',
+];
+function getAliasedActions( state, action ) {
+	if ( actionsToIgnore.includes( action.type ) ) {
+		return;
+	}
+	if ( action.clientId ) {
```

Comment:
This one handles things like updating block attributes: the action might only affect one client ID, so just apply it to the other aliased client IDs too. This however probably doesn't work for any action that might have clientId but also a root client ID 🤔 Thankfully most of those use `clientIds`.